File: AIMS-BACKEND/core/utils/payload_encryption.py
import os
import base64
import hashlib
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

DEFAULT_SECRET_KEY = "CSU_SECURE_PAYLOAD_KEY_2026_XYZ"

def _get_key_bytes() -> bytes:
    from django.conf import settings
    secret_key = getattr(settings, 'PAYLOAD_ENCRYPTION_KEY', os.getenv('PAYLOAD_ENCRYPTION_KEY', DEFAULT_SECRET_KEY))
    if not secret_key:
        secret_key = DEFAULT_SECRET_KEY
    return hashlib.sha256(secret_key.encode('utf-8')).digest()

def encrypt_payload(plaintext: str) -> str:
    if not plaintext:
        return plaintext
    
    key_bytes = _get_key_bytes()
    aesgcm = AESGCM(key_bytes)
    # Generate 12-byte nonce (IV)
    nonce = os.urandom(12)
    # Encrypt: output is Ciphertext + 16-byte Tag
    encrypted_data = aesgcm.encrypt(nonce, plaintext.encode('utf-8'), None)
    
    # Combined: [12-byte Nonce][Ciphertext][16-byte Tag]
    combined = nonce + encrypted_data
    return base64.b64encode(combined).decode('utf-8')

def decrypt_payload(ciphertext_b64: str) -> str:
    if not ciphertext_b64:
        return ciphertext_b64
    
    combined = base64.b64decode(ciphertext_b64)
    if len(combined) < 28:
        raise ValueError("Ciphertext too short (must contain 12-byte Nonce and 16-byte Tag).")
        
    nonce = combined[:12]
    encrypted_data = combined[12:] # Ciphertext + Tag
    
    key_bytes = _get_key_bytes()
    aesgcm = AESGCM(key_bytes)
    
    decrypted_data = aesgcm.decrypt(nonce, encrypted_data, None)
    return decrypted_data.decode('utf-8')

File: frontend/src/utils/payloadEncryption.ts
// AES-GCM Payload Encryption/Decryption using Web Crypto API

const DEFAULT_SECRET_KEY = import.meta.env.VITE_PAYLOAD_ENCRYPTION_KEY || "CSU_SECURE_PAYLOAD_KEY_2026_XYZ";

/**
 * Derives a raw CryptoKey from the pre-shared secret key string using SHA-256.
 */
const getRawKey = async (secret: string) => {
  const enc = new TextEncoder();
  const secretBytes = enc.encode(secret);
  const hash = await window.crypto.subtle.digest("SHA-256", secretBytes);
  return window.crypto.subtle.importKey(
    "raw",
    hash,
    { name: "AES-GCM" },
    false,
    ["encrypt", "decrypt"]
  );
};

/**
 * Encrypts standard JSON/plaintext using AES-GCM (256-bit).
 * Returns a Base64 encoded string containing [12-byte Nonce][Ciphertext][16-byte Tag].
 */
export const encryptPayload = async (plaintext: string, secret = DEFAULT_SECRET_KEY) => {
  if (!plaintext) return plaintext;
  try {
    const enc = new TextEncoder();
    const key = await getRawKey(secret);
    const iv = window.crypto.getRandomValues(new Uint8Array(12)); // 12-byte IV for GCM

    const encryptedBuffer = await window.crypto.subtle.encrypt(
      { name: "AES-GCM", iv: iv, tagLength: 128 },
      key,
      enc.encode(plaintext)
    );

    // Web Crypto API appends the 16-byte authentication tag at the end of the encrypted buffer
    const combined = new Uint8Array(iv.length + encryptedBuffer.byteLength);
    combined.set(iv, 0);
    combined.set(new Uint8Array(encryptedBuffer), iv.length);

    // Convert bytes to Base64 safely
    let binary = "";
    const len = combined.byteLength;
    for (let i = 0; i < len; i++) {
      binary += String.fromCharCode(combined[i]);
    }
    return btoa(binary);
  } catch (err) {
    console.error("[CRYPTO ERROR] Encryption failed:", err);
    throw err;
  }
};

/**
 * Decrypts a Base64 ciphertext back to standard plaintext using AES-GCM.
 */
export const decryptPayload = async (base64Ciphertext: string, secret = DEFAULT_SECRET_KEY) => {
  if (!base64Ciphertext) return base64Ciphertext;
  try {
    const key = await getRawKey(secret);
    const binary = atob(base64Ciphertext);
    const len = binary.length;
    const bytes = new Uint8Array(len);
    for (let i = 0; i < len; i++) {
      bytes[i] = binary.charCodeAt(i);
    }

    if (bytes.length < 28) {
      throw new Error("Ciphertext too short (must be at least 28 bytes containing 12-byte Nonce and 16-byte Tag).");
    }

    const iv = bytes.slice(0, 12);
    const encryptedData = bytes.slice(12); // Contains [Ciphertext][Tag]

    const decrypted = await window.crypto.subtle.decrypt(
      { name: "AES-GCM", iv: iv, tagLength: 128 },
      key,
      encryptedData
    );

    const dec = new TextDecoder();
    return dec.decode(decrypted);
  } catch (err) {
    console.error("[CRYPTO ERROR] Decryption failed:", err);
    throw err;
  }
};
